onnxruntime/core/framework/utils.h

```diff
@@ -87,69 +87,399 @@ void DumpNodeInputs(const OpKernelContext& context, const Node& node);
+  return ONNX_NAMESPACE::TensorProto_DataType_UINT64;
+};
+template <>
+constexpr ONNX_NAMESPACE::TensorProto_DataType ToTensorProtoElementType<BFloat16>() {
+  return ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16;
+};
+
+template <typename T>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<bool>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<std::string>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<float>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<double>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_DOUBLE;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<MLFloat16>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<BFloat16>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_BFLOAT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int8_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint8_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT8;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int16_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint16_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int32_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint32_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT32;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int64_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint64_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64;
+}
+
+// The following primitives are strongly recommended for switching on tensor input datatypes for
+// kernel implementations.
+//
+//  1) If you need to handle all of the primitive tensor contained datatypes, the best choice would be macros
+//     DispatchOnTensorType or DispatchOnTensorTypeWithReturn. Use inline wrappers so your function can be invoked as function<T>().
+//  2) if you have a few types, use utils::IsPrimitiveDataType<T>() for a sequence of if/else statements.
+//  3) For something in between, we suggest to use CallDispatcher pattern.
```

Comment:
Why go through the indirection of IsPrimitiveDataType<T> vs just comparing the return value to a ONNX_NAMESPACE::TensorProto_DataType_* value?

Reply:
Because ONNX_NAMESPACE::... is wordy and creates a tie to Protobuf constants. These wrappers make it easier, intent is clear and we can remove protobuf if necessary.

---
In reply to: [342909190](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r342909190) [](ancestors = 342909190)